app/src/org/commcare/tasks/DataPullTask.java

```diff
@@ -531,41 +531,35 @@ private static void recordSuccessfulSyncTime(String username) {
         //this is the temporary implementation of everything past this point
 
         //Wipe storage
-        //TODO: move table instead. Should be straightforward with sandboxed db's
-        wipeStorageForFourTwelveSync();
+        SQLiteDatabase userDb = CommCareApplication.instance().getUserDbHandle();
+        wipeStorageForFourTwelveSync(userDb);
 
         String failureReason = "";
         try {
-            //Get new data
-            String syncToken = readInput(cache.retrieveCache(), factory);
+            String syncToken = readInputWithoutCommit(cache.retrieveCache(), factory);
             updateUserSyncToken(syncToken);
             Logger.log(LogTypes.TYPE_USER, "Sync Recovery Successful");
+            userDb.setTransactionSuccessful();
             return new Pair<>(PROGRESS_DONE, "");
-        } catch (ActionableInvalidStructureException e) {
-            e.printStackTrace();
-            failureReason = e.getLocalizedMessage();
         } catch (InvalidStructureException | XmlPullParserException
-                | UnfullfilledRequirementsException | SessionUnavailableException | IOException e) {
-            e.printStackTrace();
-            failureReason = e.getMessage();
+                | UnfullfilledRequirementsException | SessionUnavailableException
+                | IOException e) {
+            Logger.exception("Sync recovery failed|" + e.getLocalizedMessage(), e);
         } finally {
             //destroy temp file
+            userDb.endTransaction();
             cache.release();
         }
-
-        //OK, so we would have returned success by now if things had worked out, which means that instead we got an error
-        //while trying to parse everything out. We need to recover from that error here and rollback the changes
-
-        //TODO: Roll back changes
-        Logger.log(LogTypes.TYPE_USER, "Sync recovery failed|" + failureReason);
         return new Pair<>(PROGRESS_RECOVERY_FAIL_BAD, failureReason);
     }
 
-    private void wipeStorageForFourTwelveSync() {
-        CommCareApplication.instance().getUserStorage(ACase.STORAGE_KEY, ACase.class).removeAll();
-        new AndroidCaseIndexTable().wipeTable();
-        CommCareApplication.instance().getUserStorage(Ledger.STORAGE_KEY, Ledger.class).removeAll();
-        EntityStorageCache.tryWipeCache();
+    private void wipeStorageForFourTwelveSync(SQLiteDatabase userDb) {
+        userDb.beginTransaction();

```

Comment:
I know it's functionally equivalent but can you move this call to the line before `wipeStorageForFourTwelveSync()` is called instead? It just improves the readability of clearly seeing the transaction start and end in a linear manner

Reply:
Yep, that's nicer